jmclient/jmclient/payjoin.py

```diff
@@ -633,5 +715,290 @@ def process_payjoin_proposal_from_server(response_body, manager):
         # if transaction is succesfully broadcast, remove the
         # timeout fallback to avoid confusing error messages:
         manager.timeout_fallback_dc.cancel()
+        manager.set_broadcast(True)
     if manager.mode == "command-line" and reactor.running:
-        reactor.stop()
+        process_shutdown()
+
+""" Receiver-specific code
+"""
+
+class PayjoinServer(Resource):
+    def __init__(self, wallet_service, mixdepth, destination,
+                 amount, mode="command-line", pj_version = 1):
+        self.pj_version = pj_version
+        self.wallet_service = wallet_service
+        self.manager = JMPayjoinManager(self.wallet_service, mixdepth,
+                                        destination, amount, mode=mode)
+        super().__init__()
+
+    isLeaf = True
+
+    def bip78_error(self, request, error_meaning,
+                    error_code="unavailable", http_code=400):
+        """
+        See https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#receivers-well-known-errors
+
+        We return, to the sender, stringified json in the body as per the above.
+        In case the error code is "original-psbt-rejected", we do not have
+        any valid payment to broadcast, so we shut down with "not paid".
+        for other cases, we schedule the fallback for 60s from now.
+        """
+
+        request.setResponseCode(http_code)
+        request.setHeader(b"content-type", b"text/html; charset=utf-8")
+        log.debug("Returning an error: " + str(
+            error_code) + ": " + str(error_meaning))
+        if error_code in ["original-psbt-rejected", "version-unsupported"]:
+            log.warn("Payment is not valid. Payment has NOT been made.")
+            # shutdown now but wait until response is sent.
+            task.deferLater(reactor, 2.0, process_shutdown)
+        else:
+            reactor.callLater(60.0, fallback_nonpayjoin_broadcast,
+                              error_meaning.encode("utf-8"), self.manager)
+        return json.dumps({"errorCode": error_code,
+                           "message": error_meaning}).encode("utf-8")
+
+    def render_GET(self, request):
+        # can be used e.g. to check if an ephemeral HS is up
+        # on Tor Browser:
+        return "<html>Only for testing.</html>".encode("utf-8")
+
+    def render_POST(self, request):
+        """ The sender will use POST to send the initial
+        payment transaction.
+        """
+        log.debug("The server got this POST request: ")
+        # unfortunately the twisted Request object is not
+        # easily serialized:
+        log.debug(request)
+        log.debug(request.method)
+        log.debug(request.uri)
+        log.debug(request.args)
+        sender_parameters = request.args
+        log.debug(request.path)
+        log.debug(request.content)
```

Comment:
This outputs "[DEBUG]  <_io.BytesIO object at 0x7f5df6443048>", that's useless, actual contents in hex should be logged I think.

Reply:
Agreed.

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/682/commits/57b07466f3ee272836f944290c0284fc201a89a7